### DISCUSSION TITLE
Starting with docker on a dev machine

### DISCUSSION THREAD

**Discussion by cadilhac** [ravendb/ravendb#19571] [Q&A] [answered]
Sorry, very newbie questions with docker/ravendb.

I am trying to setup a local container on my dev machine running under Win10. Docker desktop is installed (it uses WSL2) and a ravendb 5.4 image is used. I am new to docker. Not so with RavenDb basic installation (dev or prod).

1. If I use docker desktop to start the image,setting ports 38888 and 8080 (what is 161 used for?), then the logs tell me to connect to http://1b30270fc299:8080. It doesn't work. http://localhost:8080 works though. Why?
2. I then select New cluster / Let's Encrypt, set a community license, set a domain. Contrarily to a local installation I have on linux, I can't enter a local ip/hostname because it's docker. What should I put here? When I use your ps1 file to start a container (instead of using the app), I see that the container's ip address is 172.17.0.2. So I put this one. The let's encrypt does its thing.
3. It tells me the https://a.xxxxxxxxxxxxxxx.ravendb.community url and asks me to re
start the server. I also install the certificate.
4. I then see in the container logs that it restarted successfully and that it listens at a.xxxxxxxxxxxxxxx.ravendb.community but this page does not answer (even if adding a :8080) so I'm clueless.

Thanks for your help.

**Comment by cadilhac** [comment]
ok, I understood that there are 2 ip here. The WSL ip and the container IP. I was able to succeed with the install by configuring the ports like so:

![config](https://github.com/user-attachments/assets/8116f856-67b8-4319-bd05-6a9b132dd662)

Studio runs well. But when I create a database, I get this error:

> System.IO.InvalidDataException: The database 'test' was created but is not accessible, because one or more of the nodes on which this database was supposed to reside on, threw an exception.
 ---> Raven.Server.Exceptions.RaftIndexWaitAggregateException: An error occurred while executing the raft command (number '19'). There is one exception aggregated within this exception. (Connection refused (a.dockertest-cadilhac.ravendb.community:443))
 ---> System.Net.Http.HttpRequestException: Connection refused (a.dockertest-cadilhac.ravendb.community:443)
 ---> System.Net.Sockets.SocketException (111): Connection refused

Any idea?

**Reply by karmeli87** [reply]
Hi,

Such errors are often an indication that something with the ports forwarding is off.

How do you run the docker? you need to publish 3 ports `-p 8080:8080 -p 443:443 -p 38888:38888`

**Reply by cadilhac** [reply]
I used Docker desktop to run a container. Since the UI does not even mention 443 in the options I had no idea that I had to forward port 443:

![port443](https://github.com/user-attachments/assets/a3b669f0-6203-4cc5-a7c7-abe57761df2b)

Moreover, it's not specified [in your doc](https://ravendb.net/docs/article-page/6.2/csharp/start/installation/running-in-docker-container) either. It only shows 8080 and 38888 in the example commande lines.

Now I run:

```
docker run --rm -d -p 8080:8080 -p 443:443 -p 38888:38888 -v /var/lib/ravendb/data ravendb/ravendb:ubuntu-latest
```

and I get another kind of exception when creating a new db:

> System.IO.InvalidDataException: The database 'test' was created but is not accessible, because one or more of the nodes on which this database was supposed to reside on, threw an exception.
 ---> Raven.Server.Exceptions.RaftIndexWaitAggregateException: An error occurred while executing the raft command (number '9'). There is one exception aggregated within this exception. (The SSL connection could not be established, see inner exception.)
 ---> System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.
 ---> System.IO.IOException: Received an unexpected EOF or 0 bytes from the transport stream.

Also, strangely, inside the container, /etc/ravendb/settings.json has this line:

> "PublicServerUrl": "https://a.dockertest-cadilhac.ravendb.community:443"

Why 443? I specified 8080. And anywyay, the wizard redirected to this hostname but with port 8080 and studio appears. With 443, it does not.

Then another difficulty: I also tried to map the data folder inside my host (-v /var/lib/ravendb/data) and it does nothing. Nothing is created in my host folder. If I run the container with -v ~/ravendb/data:/var/lib/ravendb/data then I get an error:

> System.InvalidOperationException: Could not access some of the specified paths. Please check if you have sufficient privileges to access following paths:
> Key: 'DataDir' Path: '/var/lib/ravendb/data' Error: 'Access to the path '/var/lib/ravendb/data/71aa9302ace349aeadd52d5f76ad8680' is denied.'

At last I investigated how the container is launched in production. What is the way to do? Automatically start an existing container at startup? Or run only ephemeral containers with a mount to a host database? Your doc gives an example command line:

```
docker run --rm -d -p 8080:8080 -p 38888:38888 -v /var/lib/ravendb/data --name RavenDb-WithData 
-e RAVEN_Setup_Mode=None -e RAVEN_License_Eula_Accepted=true 
-e RAVEN_Security_UnsecuredAccessAllowed=PrivateNetwork ravendb/ravendb
```

But this does not seem to be for prod. With that, the original license is lost, studio does appear only with an ip, not with my community subdomain.

All this makes me say that your docker documentation is not for beginners. It's really difficult to setup a working container... After reading your doc, I should be able to issue docker commands to setup a container, either for dev (for instance local WSL / docker), or for prod (linux VPS for instance).

**Reply by karmeli87** [reply, marked as answer]
We definitely need to improve the docs around this area, please feel free to add more stuff to this issue https://issues.hibernatingrhinos.com/issue/RDoc-3102

After some trial and error this is what worked for me 
`docker run --rm -p 8080:8080 -p 38888:38888 -v ravendb-data:/var/lib/ravendb/data -v ravendb-etc:/etc/ravendb ravendb/ravendb`
We need to mount the `/etc/ravendb` as well as it contain the certificates and the `settings.json`

I've just noticed that in your case I think you need to do `-p 443:8080` to redirect the traffic from the 443 port to the 8080 port that you defined in the setup.

**Reply by cadilhac** [reply]
I tried your command line and it works well. But note 2 points:
1. I have to go to http://localhost:443 (which is strange because http and 443 seem to be opposite) to start the wizard. I wanted to use 443 so that, later, I could go to https://a.xxxxxxxxxx.ravendb.community (without the terminal :8080). And you told me to do this way to avoid the exception mentioned earlier.
2. At the end of the setup, the wizard incorrectly redirects to the community domain with a :8080. So I have to remove the port in the url and it works (uses 443 then).

Killing the container and running it again with the same command line allows to retrieve the databases created in a previous container correctly.

Of course this command line hides where the volumes really are. It would be nice to explain how to find them or how to specify local directories instead of volumes.

**Reply by cadilhac** [reply]
A question about docker compose. I'm trying to compose a web server and the ravendb server.
The ravendb part shows:

```yaml
  db:
    image: ravendb/ravendb:latest-lts
    volumes:
      - ravendb-data:/var/lib/ravendb/data
      - ravendb-etc:/etc/ravendb
    ports:
      - "38888:38888"
      - "443:8080"

  volumes:
    ravendb-data:
      external: true
    ravendb-etc:
      external: true
```

As said previously, running "docker run" works well now. But when running docker compose, ravendb throws an exception:

> In Linux low-level port (below 1024) will need a special permission, if this is your case please run
  sudo setcap CAP_NET_BIND_SERVICE=+eip /usr/lib/ravendb/server/Raven.Server
  Urls: [https://172.17.0.2:8080, tcp://172.17.0.2:38888]
  System.Net.Sockets.SocketException (99): Cannot assign requested address

Any idea what I'm doing wrong?

**Reply by cadilhac** [reply]
Adding this to the db section solves the issue, but I have no idea if this is the right way to do it:
```yaml
network_mode: "bridge"
```

When my web server and my ravendb server were run through "docker run", everything was working. But when started with docker compose, then my web server is not able to reach https://a.xxxxxxxxxxxxxx.ravendb.community (port 443):

> System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.
    System.IO.IOException: Received an unexpected EOF or 0 bytes from the transport stream.

Then, starting manually the web server in the containers stack a few seconds later works. So this is a question of timing. The web  server container should start only when the db is up and running. I finally found how to set this. In the compose file, for the web server:

```yaml
depends_on:
  db:
    condition: service_healthy
```

As you can see, this is this kind of things that a complete guide could solve.

**Reply by poissoncorp** [reply]
The problem is that you're using `ravendb:latest-lts`, which is currently 6.2. Starting from RavenDB 6.0, Docker images run without root permissions for security. This means the container cannot bind to low-level ports (below 1024, e.g., port 443) by default.

To enable this without running the container as root, you can add:

```
cap_add:
  - NET_BIND_SERVICE
```

This grants the container the capability to bind to low-level ports without requiring root. The capability is disabled by default in Docker for security reasons.

WSL runs processes inside a virtualized environment with a shared network bridge that maps container traffic to the Windows network stack. When you set network_mode: "bridge", the container operates with an isolated networking stack. In this mode, the container binds the port internally to its own stack, and Docker uses NAT to map traffic, bypassing the host-level port binding restrictions imposed by WSL.

However, note that the user-defined bridge network (default in Docker Compose) includes automatic DNS resolution, allowing services to communicate by name. Explicitly setting network_mode: "bridge" disables this feature, which might cause connectivity issues if your services rely on Docker's DNS.

Using cap_add: NET_BIND_SERVICE is a more consistent and universal solution. It works regardless of the host system (Linux, WSL, Mac) while retaining Docker's modern networking features, such as DNS-based service discovery.

**Reply by cadilhac** [reply]
Thanks for the information. But when removing the bridge mode and adding the capability, it does not work.

Also, I found [this paper](https://medium.com/@olivier.gaumond/why-am-i-able-to-bind-a-privileged-port-in-my-container-without-the-net-bind-service-capability-60972a4d5496) saying:

> It turns out that [was changed in Docker](https://github.com/moby/moby/pull/41030) since version 20.03. Basically they defined that unprivileged port start at 0 instead of 1024 so the NET_BIND_SERVICE capability is not needed

**Reply by poissoncorp** [reply]
I see that in this scenario we have two convoluted issues here:
1. unable to start the RavenDB process
2. web server is not able to reach RavenDB by the https://a.xxxxxxxxxxxxxx.ravendb.community/ name

To break this down, could you exec bash on your working web server while in "bridge" mode to address the connectivity issues? Use dig to check whether the domain resolves correctly

**Reply by cadilhac** [reply]
Not sure what you want me to do. In bridge mode, I can exec a shell (sh since bash is not in the image). dig is not available. I can ping my community subdomain.

**Comment by cadilhac** [comment]
@karmeli87 Any ETA for this doc? I'm struggling to setup ravendb on a docker swarm in prod...

**Reply by karmeli87** [reply]
Hi, 

I know that we are currently working on it, but unfortunately I don't have an ETA.

I understand that you are experience difficulties on setting up the ravendb cluster, please feel free to share what you did and what is not working. 

We will try to provide you with a solution, regardless to the documentation task :-)

**Reply by cadilhac** [reply]
While trying things, is it a good idea to setup containerized RavenDb (unprotected) behind the traefik reverse proxy that can handle let's encrypt certificates automatically?

**Reply by ayende** [reply]
No, RavenDB needs to do it's own SSL termingation

**Reply by cadilhac** [reply]
ok, so will try to pass HTTPS through traefik to reach ravendb. I think I have to do that because traefik will also handle a web site on :443 so I can't let ravendb also listen on :443 by itself.

**Reply by georgiosd** [reply]
@cadilhac it's been discussed many times here that this is a bad idea. RavenDB does some pretty unique things and Traefik and other proxies are designed for "normal" web services. 

Just change the HTTPS port of RavenDB to another port and leave 443 for web?

**Reply by cadilhac** [reply]
I hear you. Thanks for pointing me in the right direction.

**Comment by cadilhac** [comment]
In prod, I still get an error in Studio after the whole setup. Studio seems to work fine. I can create a database. When I want to delete it, the db greys out (says "loading" on its left side) and I get this error message in the notif center:

> Topology of database 'test' was changed
> December 18, 2024 9:38 AM
> Node A is currently not responding (with status: None) and moved to rehab (739237.14:38:59.3780647)

I use docker stack to publish my stack. Here is the ravendb part:

```
  db:
    image: ravendb/ravendb:latest-lts
    volumes:
      - ravendb-data:/var/lib/ravendb/data
      - ravendb-etc:/etc/ravendb
    cap_add:
      - NET_BIND_SERVICE # just in case when I change 7012:8080 to 7012:443 to see if it fixes the issue
    ports:
      - "38888:38888"
      - "7012:8080"
      - "8081:8080"
```

After setup, the resulting config file is:

```
{
  "License.Eula.Accepted": true,
  "Security.Certificate.LetsEncrypt.Email": "........@gmail.com",
  "Setup.Mode": "LetsEncrypt",
  "Security.Certificate.Path": "/etc/ravendb/security/server.pfx",
  "ServerUrl": "https://0.0.0.0:8080",
  "ServerUrl.Tcp": "tcp://0.0.0.0:38888",
  "ExternalIp": "<my VPS ip>",
  "PublicServerUrl": "https://a.mydomain.ravendb.community:7012",
  "PublicServerUrl.Tcp": "tcp://a.mydomain.ravendb.community:38888"
}
```

I use 0.0.0.0 in the ServerUrl because it's a docker container. Each time it restarts, the internal ip is different.

When I run the setup, I go to http://mydomain:8081.
When it's finished and the server restarts, it goes to https://a.mydomain.ravendb.community:7012. Studio appears right away. And then I get the bug I described above.

**Reply by cadilhac** [reply]
I can't continue my tests. I get this error during setup:

```
[16:00:03 INFO] Setting up RavenDB in Let's Encrypt security mode.
[16:00:03 INFO] Getting challenge(s) from Let's Encrypt. Using e-mail: .......@gmail.com.
[16:02:04 ERROR] Setting up RavenDB in Let's Encrypt security mode failed.
Setting up RavenDB in Let's Encrypt security mode failed.
System.InvalidOperationException: Setting up RavenDB in Let's Encrypt security mode failed.
 ---> System.InvalidOperationException: Failed to receive challenge(s) information from Let's Encrypt.
 ---> System.InvalidOperationException: Let's Encrypt client failed to send the request (with retries): Method: POST, RequestUri: 'https://acme-v02.api.letsencrypt.org/acme/new-order', Version: 1.1, Content: System.Net.Http.StringContent, Headers:
{
  User-Agent: RavenDB/6.2
  User-Agent: (Ubuntu 22.04.5 LTS;X64;.NET 8.0.11;X64;;;6.2.2)
  Content-Type: application/jose+json
  Content-Length: 1356
}
 ---> System.InvalidOperationException: The request message was already sent. Cannot send the same request message multiple times.
```

Is it because I have done too many tries and Let'sEncrypt stops responding during a few hours? Is there no way to RavenDb to store the SSL cert in a mounted volume so that a next installation would reuse the still valid cert?

**Reply by ayende** [reply]
The likely reason is that the _container_ isn't able to access itself.
It is trying to ping the public endpoint, and cannot.
Can you setup so it would be accessible?

Check both ports and DNS please

**Reply by cadilhac** [reply]
From inside the container, I am able to curl https://a.mydomain.ravendb.community:7012.
What else should I check?

In the server logs, I also see this:

> 2024-12-19T13:24:36.1359688Z, 33, Information, Heartbeats supervisor from A to A in term 2, Raven.Server.ServerWide.Maintenance.ClusterMaintenanceSupervisor+ClusterNode, Exception was thrown while collecting info from A, EXCEPTION: System.Net.Sockets.SocketException (110): Connection timed out
>    at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
>    at System.Net.Sockets.Socket.<ConnectAsync>g__WaitForConnectWithCancellation|285_0(AwaitableSocketAsyncEventArgs saea, ValueTask connectTask, CancellationToken cancellationToken)
>    at System.Net.Sockets.TcpClient.CompleteConnectAsync(ValueTask task)
>    at Raven.Client.Util.TcpUtils.ConnectAsync(String url, Nullable`1 timeout, Boolean useIPv6, CancellationToken token) in /_/62011/src/Raven.Client/Util/TcpUtils.cs:line 138
>    at Raven.Client.Util.TcpUtils.ConnectSecuredTcpSocket(TcpConnectionInfo info, X509Certificate2 cert, CipherSuitesPolicy cipherSuitesPolicy, OperationTypes operationType, NegotiationCallback negotiationCallback, JsonOperationContext negContext, Nullable`1 timeout, List`1 logs, CancellationToken token) in /_/62011/src/Raven.Client/Util/TcpUtils.cs:line 312
>    at Raven.Client.Util.AsyncHelpers.HandleException(AggregateException ex, Stopwatch sw) in /_/62011/src/Raven.Client/Util/AsyncHelpers.cs:line 213
>    at Raven.Client.Util.AsyncHelpers.RunSync[T](Func`1 task) in /_/62011/src/Raven.Client/Util/AsyncHelpers.cs:line 103
>    at Raven.Server.ServerWide.Maintenance.ClusterMaintenanceSupervisor.ClusterNode.ConnectToClientNode(TcpConnectionInfo tcpConnectionInfo, TimeSpan timeout) in /_/62011/src/Raven.Server/ServerWide/Maintenance/ClusterMaintenanceSupervisor.cs:line 479
>    at Raven.Server.ServerWide.Maintenance.ClusterMaintenanceSupervisor.ClusterNode.ListenToMaintenanceWorker() in /_/62011/src/Raven.Server/ServerWide/Maintenance/ClusterMaintenanceSupervisor.cs:line 281

but unfortunately, I'm not seeing how it tries to connect to itself (what address, ip, port)...

**Reply by poissoncorp** [reply]
What about your ServerUrl (not public)? Can you curl the server within the internal network from inside the container?
If your setup involves multiple nodes, check also the connectivity between them.

**Reply by cadilhac** [reply]
As you can see, my ServerUrl is https://0.0.0.0:8080, so what address do you want me to curl exactly?

**Reply by gregolsky** [reply]
Can you try setting `ServerUrl.Cluster` to `https://${YOUR_CONTAINER_PRIVATE_IP}` please?
Reference: https://ravendb.net/docs/article-page/6.2/csharp/server/configuration/core-configuration#serverurl.cluster

**Reply by poissoncorp** [reply]
> As you can see, my ServerUrl is https://0.0.0.0:8080, so what address do you want me to curl exactly?

RavenDB Server - we need to make sure that traffic reaches RavenDB server, living inside the container (the container has internal network IP). Make sure you've established connection with RavenDB Server, when _curling_ public server URL. In addition, just to make sure, curl from inside the container - in orchestrated environment it's rare (not impossible), but worth checking that the traffic isn't blocked for some reason.

**Reply by cadilhac** [reply]
If I use 'docker inspect', I can find the internal ip of the container. Then I curl this ip:8080 (from inside the container with an 'exec bash') and it logically complains about the certificate:

> curl: (60) SSL: no alternative certificate subject name matches target host name '10.0.1.8'
> More details here: https://curl.se/docs/sslcerts.html
> 
> curl failed to verify the legitimacy of the server and therefore could not
> establish a secure connection to it. To learn more about this situation and
> how to fix it, please visit the web page mentioned above.

With -k param, it goes ok:

`curl https://10.0.1.8:8080 -k -I`

> HTTP/2 400
> content-type: application/json; charset=utf-8
> date: Fri, 20 Dec 2024 13:49:31 GMT

**Reply by cadilhac** [reply]
What is strange is that a few seconds after starting the container, I can create a db, add a document and then delete the db successfully. Then I quickly get this message in the notification center:

> Topology of database 'test' was changed
> Node A is currently not responding (with status: None) and moved to rehab (739239.14:14:49.5422232)

After that, I can't delete a db (or maybe it's deleted but it continues to appear in the db list, with a grey loading on its left side) and there are connection errors displayed in the server log as I described.

**Reply by cadilhac** [reply]
> Can you try setting `ServerUrl.Cluster` to `https://${YOUR_CONTAINER_PRIVATE_IP}` please? Reference: https://ravendb.net/docs/article-page/6.2/csharp/server/configuration/core-configuration#serverurl.cluster

@gregolsky I tried that by modifying the settings.json file inside the container. But then I have to restart the server, which reacreates the container and gives it a new ip adress which does not correspond to the one in the settings...

**Reply by cadilhac** [reply]
@gregolsky I don't know if this is the way to do it, but finally I added a hostname property inside the compose file for the ravendb service (set to ravendb), and I used https://ravendb:8080 for ServerUrl.Cluster. I checked that this hostname was understood by the container and it was. Unfortunately it changes nothing to the issue.

**Reply by cadilhac** [reply]
I discovered a page to gather debug info in Studio. Found a file named admin.node.ping.json. There I read:

> Failed to connect to all urls tcp://a.mydomain.ravendb.community:38888

Checked the VPS firewall. Port 38888 was missing in the allowed ports !! I was so obsessed by the docker and ravendb configs that I forgot to double check the VPS. Yes, I know, this was obvious. Always check the RavenDb needed ports. Forgot one the second time I created the VM...

Thanks to all those trying to help. I am now continuing my journey. See you for the next headache... ;)

**Comment by cadilhac** [comment]
This morning I met another problem and I realized something else might be worth documenting when using docker in dev env. The a.mydomain.ravendb.community hostname must point to the WSL2 ip address. This works, but what after a reboot? WSL2 ip address is ephemeral and therefore the DNS resolution fails. My workaround is to go to https://customers.ravendb.net/dns, change the ip and wait for the propagation. Any other id to make this more stable?

**Reply by ayende** [reply]
https://gist.github.com/wllmsash/1636b86eed45e4024fb9b7ecd25378ce

**Reply by cadilhac** [reply]
I tried the script at the top of your link. Unfortunately, it does not work. After running the script (with no errors), I can see the requested ip when typing "ip addr" in WSL bash. But I can't ping my host machine from WSL, and I can't ping WSL from my windows host using the new ip.

I run Windows 10.

**Comment by georgiosd** [comment]
Hey @cadilhac, happy new year :) 

Super curious here - why are you trying to create a publicly accessible cluster from your local dev environment?

**Reply by cadilhac** [reply]
Happy new year to you too!

What makes you think the db is available to outside?

**Reply by georgiosd** [reply]
Hmmm, I guess because you are trying to use a public domain. What is it you would like to achieve?

**Reply by cadilhac** [reply]
Simply setting up a good dev env and understanding what will be my prod setup on a VPS using docker (using swarm).

**Reply by georgiosd** [reply]
Hm.... in the scenario where you have to traverse NAT and route the IP to WSL is going to be much more difficult than setting up a cluster in production. You can definitely setup a local cluster to run tests with, but I would opt for a self-signed certificate and local resolution.

Re hosting in swarm - are you sure that's a good idea? Raven requires that both the nodes and the clients can speak to all nodes without any intermediate proxy. Swarm will redirect traffic from one node to another on its own criteria and that would be interference. 

Definitely not a swarm expert though, so if there is a way to do this, I'm all ears :)
